pkg/storegateway/bucket_index_postings.go

```diff
@@ -19,64 +19,126 @@ import (
+		}
 	}
+
+	return postingGroup{
+		isSubtract: g.isSubtract,
+		keys:       keys,
+	}, nil
 }
 
-// NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
-func toPostingGroup(lvr indexheader.Reader, m *labels.Matcher) (*postingGroup, error) {
+func (g rawPostingGroup) filterKeys(r indexheader.Reader) ([]labels.Label, error) {
+	keys := g.keys
+	existingKeys := 0
+	for i, l := range keys {
+		if _, err := r.PostingsOffset(l.Name, l.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
+			// This label name and value doesn't exist in this block, so there are 0 postings we can match.
+			// Set it to an empty value, so we can filter it out later.
+			keys[i] = labels.Label{}
+			// Try with the rest of the set matchers, maybe they can match some series.
+			continue
+		} else if err != nil {
+
+			return nil, err
+		}
+		existingKeys++
+	}
+	if existingKeys == len(keys) {
+		return keys, nil
+	}
+	if existingKeys == 0 {
+		return nil, nil
+	}
+
+	filtered := make([]labels.Label, 0, existingKeys)
+	for _, k := range keys {
+		if k == (labels.Label{}) {
+			continue
+		}
+		filtered = append(filtered, k)
+	}
+	return filtered, nil
+}
+
+// toPostingGroup returns a rawPostingGroup. toPostingGroup does not guarantee that
+// they keys of each rawPostingGroup exist in the index of the block. To verify this,
+// call rawPostingGroup.prepare() with an index reader.
+// NOTE: Derived from tsdb.postingsForMatcher
+func toPostingGroup(m *labels.Matcher) rawPostingGroup {
 	if setMatches := m.SetMatches(); len(setMatches) > 0 && (m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp) {
 		keys := make([]labels.Label, 0, len(setMatches))
 		for _, val := range setMatches {
-			if _, err := lvr.PostingsOffset(m.Name, val); errors.Is(err, indexheader.NotFoundRangeErr) {
-				// This label name and value doesn't exist in this block, so there are 0 postings we can match.
-				// Try with the rest of the set matchers, maybe they can match some series.
-				continue
-			}
 			keys = append(keys, labels.Label{Name: m.Name, Value: val})
 		}
 		if m.Type == labels.MatchNotRegexp {
-			return newPostingGroup(true, nil, keys), nil
+			return newRawPostingGroup(true, m.Name, keys)
 		}
-		return newPostingGroup(false, keys, nil), nil
+		return newRawPostingGroup(false, m.Name, keys)
 	}
 
 	if m.Value != "" {
 		// Fast-path for equal matching.
 		// Works for every case except for `foo=""`, which is a special case, see below.
 		if m.Type == labels.MatchEqual {
-			if _, err := lvr.PostingsOffset(m.Name, m.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
-				// This label name or value doesn't exist in this block, so there are 0 postings we can match.
-				return newPostingGroup(false, nil, nil), nil
-			}
-			return newPostingGroup(false, []labels.Label{{Name: m.Name, Value: m.Value}}, nil), nil
+			return newRawPostingGroup(false, m.Name, []labels.Label{{Name: m.Name, Value: m.Value}})
```

Comment:
Could we just pass the list of label values instead of `[]labels.Label`? Would we get any memory reduction benefit?

Reply:
maybe. we would reduce allocations until we have to construct a `[]labels.Label`. The constructed `[]labels.Labels` will be only the filtered values, so will be fewer than the number of keys we have here. I will try this

Reply:
This is the benchmark for this diff 08cb1ae01b545d632bf0e6f175ed445ba140d3cc...02a6dc24cc5c87f4de44519a68ed9afeacb10a5c


<details>
<summary>BucketIndexReader_ExpandedPostings</summary>

```
name                                                                        old time/op    new time/op    delta
BucketIndexReader_ExpandedPostings/i=~"[0-2]xxx"-10                           55.3µs ± 2%    74.4µs ±10%  +34.42%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"0xxx|1xxx|2xxx"-10                     54.9µs ± 1%    66.9µs ±17%  +21.75%  (p=0.016 n=4+5)
BucketIndexReader_ExpandedPostings/i=~".+"-10                                  274ms ± 0%     330ms ±26%  +20.52%  (p=0.032 n=5+5)
BucketIndexReader_ExpandedPostings/n="1"-10                                   4.82ms ± 2%    5.06ms ± 2%   +4.81%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",j!="foo"-10                          1.43µs ± 1%    1.49µs ± 3%   +4.33%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".*"-10                                 85.0ms ± 1%    87.9ms ± 1%   +3.40%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",j="foo"-10                           1.43µs ± 1%    1.48µs ± 1%   +3.36%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X"-10                                   1.01µs ± 0%    1.05µs ± 2%   +3.34%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10    1.77µs ± 0%    1.83µs ± 1%   +3.03%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-10           191ms ± 1%     197ms ± 1%   +2.95%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"^.+$",j=~"X.+"-10                      2.97ms ± 0%    3.06ms ± 1%   +2.92%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-10                          35.6ms ± 2%    36.5ms ± 1%   +2.73%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~""-10                                    391ms ± 0%     401ms ± 0%   +2.61%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-10                           45.0ms ± 0%    46.2ms ± 1%   +2.59%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-10             177ms ± 1%     182ms ± 1%   +2.57%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/j=~"X.+"-10                                1.31µs ± 1%    1.34µs ± 1%   +2.25%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-10                  26.3ms ± 0%    26.7ms ± 1%   +1.78%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",j="foo"-10                           45.5ms ± 3%    46.2ms ± 0%     ~     (p=0.151 n=5+5)
BucketIndexReader_ExpandedPostings/i!=""-10                                    322ms ±22%     279ms ± 1%     ~     (p=0.190 n=5+4)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-10                   49.2ms ± 3%    48.3ms ± 2%     ~     (p=0.095 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",j="foo"-10                 1.66µs ± 4%    1.85µs ±27%     ~     (p=0.056 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-10            62.5ms ±15%    60.7ms ± 6%     ~     (p=0.841 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!=""-10                              149ms ± 1%     149ms ± 2%     ~     (p=0.841 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-10                      171ms ± 1%     171ms ± 1%     ~     (p=0.548 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"X.+"-10                    2.37ms ± 3%    2.40ms ± 3%     ~     (p=0.690 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-10                    176ms ±12%     170ms ± 1%     ~     (p=0.690 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"(0|1|2)xxx"-10                         56.0µs ± 5%    70.8µs ±21%     ~     (p=0.056 n=5+5)
BucketIndexReader_ExpandedPostings/i!~[0-2]xxx-10                             96.4ms ± 1%    97.1ms ± 1%     ~     (p=0.151 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".*",_i!~[0-2]xxx-10                    98.0ms ± 1%   102.4ms ±11%     ~     (p=0.056 n=5+5)
BucketIndexReader_ExpandedPostings/p!=""-10                                   19.9ms ± 2%    19.8ms ± 1%     ~     (p=0.310 n=5+5)
BucketIndexReader_ExpandedPostings/j=~"XXX|YYY"-10                            1.05µs ± 1%    1.03µs ± 1%   -1.53%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"XXX|YYY"-10                2.01µs ± 1%    1.96µs ± 0%   -2.18%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"X|Y|Z",j="foo"-10                2.51µs ± 9%    2.21µs ± 1%  -11.92%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"X|Y|Z"-10                              1.38µs ± 0%    1.08µs ± 1%  -22.04%  (p=0.008 n=5+5)

name                                                                        old alloc/op   new alloc/op   delta
BucketIndexReader_ExpandedPostings/n="1",j="foo"-10                           19.8MB ± 0%    19.8MB ± 0%   +0.00%  (p=0.016 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"X.+"-10                    4.81MB ± 0%    4.81MB ± 0%   +0.00%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-10           100MB ± 0%     100MB ± 0%   +0.00%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-10                   93.5MB ± 0%    93.5MB ± 0%   +0.00%  (p=0.032 n=4+5)
BucketIndexReader_ExpandedPostings/n="X",j="foo"-10                           1.31kB ± 0%    1.31kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/n="X",j!="foo"-10                          1.31kB ± 0%    1.31kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/j=~"X.+"-10                                1.93kB ± 0%    1.93kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/n="1"-10                                   12.7MB ± 0%    12.7MB ± 0%     ~     (p=0.056 n=5+5)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-10                           19.8MB ± 0%    19.8MB ± 0%     ~     (p=1.000 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-10                          19.8MB ± 0%    19.8MB ± 0%     ~     (p=0.794 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".*"-10                                  287MB ± 0%     287MB ± 0%     ~     (p=0.452 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".+"-10                                  327MB ± 0%     327MB ± 0%     ~     (p=0.571 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"^.+$",j=~"X.+"-10                      4.81MB ± 0%    4.81MB ± 0%     ~     (p=0.333 n=5+5)
BucketIndexReader_ExpandedPostings/i=~""-10                                    113MB ± 0%     113MB ± 0%     ~     (p=0.357 n=5+5)
BucketIndexReader_ExpandedPostings/i!=""-10                                    327MB ± 0%     327MB ± 0%     ~     (p=0.667 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-10                   21.4MB ± 0%    21.4MB ± 0%     ~     (p=0.095 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",j="foo"-10                 1.49kB ± 0%    1.49kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-10            22.7MB ± 0%    22.7MB ± 0%     ~     (p=0.151 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!=""-10                             86.4MB ± 0%    86.4MB ± 0%     ~     (p=0.198 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-10                     93.5MB ± 0%    93.5MB ± 0%     ~     (p=0.595 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-10                  23.9MB ± 0%    23.9MB ± 0%     ~     (p=0.167 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-10            94.8MB ± 0%    94.8MB ± 0%     ~     (p=0.635 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10    1.67kB ± 0%    1.67kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"0xxx|1xxx|2xxx"-10                     14.2kB ± 0%    14.2kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"(0|1|2)xxx"-10                         14.1kB ± 0%    14.1kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"[0-2]xxx"-10                           14.1kB ± 0%    14.1kB ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i!~[0-2]xxx-10                              285MB ± 0%     285MB ± 0%     ~     (p=0.937 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".*",_i!~[0-2]xxx-10                     287MB ± 0%     287MB ± 0%     ~     (p=0.706 n=5+5)
BucketIndexReader_ExpandedPostings/p!=""-10                                   58.8MB ± 0%    58.8MB ± 0%     ~     (p=0.421 n=5+5)
BucketIndexReader_ExpandedPostings/n="X"-10                                     977B ± 0%      961B ± 0%   -1.64%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"XXX|YYY"-10                1.52kB ± 0%    1.47kB ± 0%   -3.16%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"X|Y|Z",j="foo"-10                1.62kB ± 0%    1.55kB ± 0%   -3.96%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/j=~"XXX|YYY"-10                              992B ± 0%      928B ± 0%   -6.45%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/i=~"X|Y|Z"-10                              1.04kB ± 0%    0.94kB ± 0%   -9.23%  (p=0.008 n=5+5)

name                                                                        old allocs/op  new allocs/op  delta
BucketIndexReader_ExpandedPostings/n="X",j="foo"-10                             28.0 ± 0%      29.0 ± 0%   +3.57%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",j!="foo"-10                            28.0 ± 0%      29.0 ± 0%   +3.57%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",j="foo"-10                   30.0 ± 0%      31.0 ± 0%   +3.33%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"X.+"-10                      31.0 ± 0%      32.0 ± 0%   +3.23%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10      31.0 ± 0%      32.0 ± 0%   +3.23%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"X|Y|Z",j="foo"-10                  33.0 ± 0%      34.0 ± 0%   +3.03%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",i!="2",j="foo"-10               149 ± 0%       153 ± 0%   +2.68%  (p=0.029 n=4+4)
BucketIndexReader_ExpandedPostings/n="1",j="foo"-10                              121 ± 0%       124 ± 0%   +2.48%  (p=0.016 n=5+4)
BucketIndexReader_ExpandedPostings/n="1",i=~".*",j="foo"-10                      124 ± 0%       126 ± 0%   +1.61%  (p=0.016 n=4+5)
BucketIndexReader_ExpandedPostings/n="1",j!="foo"-10                             120 ± 1%       121 ± 0%   +1.17%  (p=0.016 n=5+4)
BucketIndexReader_ExpandedPostings/j="foo",n="1"-10                              122 ± 0%       123 ± 0%   +1.15%  (p=0.016 n=5+4)
BucketIndexReader_ExpandedPostings/n="1"-10                                     93.0 ± 0%      94.0 ± 0%   +1.08%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~"1.+",j="foo"-10                   66.8k ± 0%     66.8k ± 0%   +0.00%  (p=0.016 n=5+4)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!="2",j="foo"-10              600k ± 0%      600k ± 0%   +0.00%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",i!~"2.*",j="foo"-10            667k ± 0%      667k ± 0%   +0.00%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i=~".+",j="foo"-10                     600k ± 0%      600k ± 0%   +0.00%  (p=0.016 n=5+4)
BucketIndexReader_ExpandedPostings/n="1",i!=""-10                               600k ± 0%      600k ± 0%   +0.00%  (p=0.016 n=4+5)
BucketIndexReader_ExpandedPostings/n="X"-10                                     23.0 ± 0%      23.0 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/j=~"X.+"-10                                  23.0 ± 0%      23.0 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~".*"-10                                    108 ± 1%       108 ± 0%     ~     (p=0.444 n=5+5)
BucketIndexReader_ExpandedPostings/i=~".+"-10                                   600k ± 0%      600k ± 0%     ~     (p=1.000 n=4+5)
BucketIndexReader_ExpandedPostings/i=~"^.+$",j=~"X.+"-10                        28.0 ± 0%      28.0 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~""-10                                     600k ± 0%      600k ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i!=""-10                                     600k ± 0%      600k ± 0%     ~     (p=0.556 n=4+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j="foo"-10                       600k ± 0%      600k ± 0%     ~     (p=0.238 n=5+5)
BucketIndexReader_ExpandedPostings/n="1",i!="",j=~"XXX|YYY"-10                  30.0 ± 0%      30.0 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"0xxx|1xxx|2xxx"-10                        118 ± 0%       118 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"(0|1|2)xxx"-10                            118 ± 0%       118 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"[0-2]xxx"-10                              118 ± 0%       118 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i!~[0-2]xxx-10                                166 ± 0%       166 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~".*",_i!~[0-2]xxx-10                       171 ± 0%       171 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/p!=""-10                                     99.0 ± 0%      99.0 ± 0%     ~     (all equal)
BucketIndexReader_ExpandedPostings/i=~"X|Y|Z"-10                                25.0 ± 0%      24.0 ± 0%   -4.00%  (p=0.008 n=5+5)
BucketIndexReader_ExpandedPostings/j=~"XXX|YYY"-10                              24.0 ± 0%      23.0 ± 0%   -4.17%  (p=0.008 n=5+5)
```

</details>

I think the gains in memory may be countered by slower processing times. The memory gains seem to be on already low-memory test cases, but the increase in CPU seems to be across both heavy and light test cases. I think we should not include this

Reply:
omitting this copy also caused a data race - see https://github.com/grafana/mimir/pull/3872#issuecomment-1382411067

(thank God for the golang race detector)

Reply:
Ok, thanks for checking it anyway!